### ISSUE TITLE
New bob learning rate schedule implementation in python training scripts.

### ISSUE THREAD

**Issue by vijayaditya** [kaldi-asr/kaldi#577]
I am writing this down in order to track the progress of this effort over the next few months. This is a low-priority so the progress would be slow. This has some overlap with #538 .

We would like to implement the new bob learning rate schedule in the training scripts. Major change required would be ensuring that the compute_{train,valid}_prob.sh are synchronized with the training jobs. We would have to wait for these jobs to complete to decide the learning rate for the next iteration.

As the compute jobs are usually not very expensive e.g. the size of the data is equivalent to 8 minibatches, it might be really cheap to run these on the GPU. However these jobs use the averaged/selected model, so these jobs cannot be run with the training jobs. This means that each iteration would now require an additional request for the GPU, which could increase training time significantly on busy clusters. 

This will be the modification compared to standard implementation: We will always use objfn values from an older iteration (iter-1) to decide the learning rate for the next iteration (iter+1). This will mean that we can combine the compute*prob.sh jobs of iter-1 with the training jobs  at current iteration (iter). I will add these commands to training job 0 in every iteration.

I will try to implement this as an option in the current training scripts, as all that has to be changed is \ TrainNewModels function. I just plan to add a new function TrainNewModelsNewbobLr, which will return the learning rate for the next iteration when the current iteration is done.

@danpovey @jtrmal please tag as an enhancement.

**Comment by danpovey** [comment]
It might be necessary to wait for even older objfn values, like iter-2 (or
make it configurable), as with modern GPUs and depending on the amount of
validation data, it could take more than twice as long to compute the
validation accuracy, as do a training iteration.
I think it might make sense to, once you decide to reduce the learning
rate, wait another iteration and check that the decision is still
consistent (i.e. objf still worse than the threshold), as otherwise we
could be misled by occasional blips.

Dan

**Comment by danpovey** [comment]
No, just use the new learning rate for the next iteration.

**Comment by danpovey** [comment]
BTW, I think left and right context are confusing terms.  Maybe talk about
'num-iterations-delay' and 'num-iterations-confirm' or something like that.

**Comment by jfainberg** [comment]
Hi @vijayaditya ,

I was wondering whether you have this implemented in one of your branches? (I couldn't find it). I need to implement it for one of my experiments.

Cheers,
Joachim

**Comment by vijayaditya** [comment]
I didn't get a chance to do it.

Vijay

On Jan 16, 2017 10:35 AM, "Joachim" <notifications@github.com> wrote:

Hi @vijayaditya <https://github.com/vijayaditya> ,

I was wondering whether you have this implemented in one of your branches?
(I couldn't find it). I need to implement it for one of my experiments.

Cheers,
Joachim

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/kaldi-asr/kaldi/issues/577#issuecomment-272893394>, or mute
the thread
